pkgs/development/libraries/gsm/default.nix

```diff
@@ -1,41 +1,53 @@
-{ stdenv, fetchurl }:
+{ stdenv, fetchurl
+, staticSupport ? false # Compile statically (support for packages that look for the static object)
+}:
+
+let
+  inherit (stdenv) isCygwin isDarwin;
+  inherit (stdenv.lib) optional;
+in
 
-with stdenv.lib;
 stdenv.mkDerivation rec {
   name = "gsm-${version}";
-  version = "1.0.13";
+  version = "1.0.14";
 
   src = fetchurl {
     url = "http://www.quut.com/gsm/${name}.tar.gz";
-    sha256 = "1bcjl2h60gvr1dc5a963h3vnz9zl6n8qrfa3qmb2x3229lj1iiaj";
+    sha256 = "0b1mx69jq88wva3wk0hi6fcl5a52qhnq2f9p3f3jdh5k61ma252q";
   };
 
-  patchPhase = ''
+  patchPhase = [ ''
     # Fix include directory
     sed -e 's,$(GSM_INSTALL_ROOT)/inc,$(GSM_INSTALL_ROOT)/include/gsm,' -i Makefile
-
-    makeFlags="$makeFlags INSTALL_ROOT=$out"
-
-    # Build shared library instead of static
-    sed -e 's,-c -O2 -DNeedFunctionPrototypes=1,-c -O2 -fPIC -DNeedFunctionPrototypes=1,' -i Makefile
-    sed -e 's,libgsm.a,libgsm.so,' -i Makefile
-    sed -e 's/$(AR) $(ARFLAGS) $(LIBGSM) $(GSM_OBJECTS)/$(LD) -shared -Wl,-soname,libgsm.so -o $(LIBGSM) $(GSM_OBJECTS) -lc/' -i Makefile
-    sed -e 's,$(RANLIB) $(LIBGSM),,' -i Makefile
-  '';
+  '' ] ++ (if staticSupport then [ ] else
+    (if isDarwin then [ ''
+      # Build dylib on Darwin
+      sed -e 's,libgsm.a,libgsm.dylib,' -i Makefile
+      sed -e 's,$(AR) $(ARFLAGS) $(LIBGSM) $(GSM_OBJECTS),$(LD) -o $(LIBGSM) -dynamiclib -install_name $(GSM_INSTALL_ROOT)/$(LIBGSM) $(GSM_OBJECTS) -lc,' -i Makefile
+    '' ] else [ ''
+      # Build ELF shared object by default
+      sed -e 's,libgsm.a,libgsm.so,' -i Makefile
+      sed -e 's/$(AR) $(ARFLAGS) $(LIBGSM) $(GSM_OBJECTS)/$(LD) -shared -Wl,-soname,libgsm.so -o $(LIBGSM) $(GSM_OBJECTS) -lc/' -i Makefile
+    '' ] ++ [ ''
+      # Hack to add cflags for shared libraries
+      sed -e 's,-c -O2 -DNeedFunctionPrototypes=1,-c -O2 -DNeedFunctionPrototypes=1 ${NIX_CFLAGS_COMPILE},' -i Makefile
+      # Remove line that is unused when building shared libraries
+      sed -e 's,$(RANLIB) $(LIBGSM),,' -i Makefile
```

Comment:
Yeah, normally I would catch a silly mistake like that, but since I can't test on darwin I didn't notice.  Also moving the closing parenthese would fix it too, but I like the optionalString method so I changed it around a bit to use that.